redfish-core/include/utils/pcie_util.hpp

```diff
@@ -24,27 +24,90 @@ namespace pcie_util
 {
 
 /**
- * @brief Populate the PCIe Device list from a GetSubTreePaths search of
- *        inventory
+ * @brief Workaround to handle duplicate PCI device list
  *
- * @param[i,o] asyncResp  Async response object
- * @param[i]   Name   Key to store the list of PCIe devices in asyncResp
+ * retrieve PCI device endpoint information and if path is
+ * ~/chassisN/io_moduleN/slotN/adapterN then, replace redfish
+ * PCI device as "chassisN_io_moduleN_slotN_adapterN"
  *
- * @return void
+ * @param[i]   fullPath  object path of PCIe device
+ *
+ * @return string: unique PCIe device name
  */
+inline std::string buildPCIeUniquePath(const std::string& fullPath)
+{
+    sdbusplus::message::object_path path(fullPath);
+    sdbusplus::message::object_path parentPath = path.parent_path();
+    sdbusplus::message::object_path parentofParentPath =
+        parentPath.parent_path();
+    std::string devName;
+
+    if (!parentofParentPath.parent_path().filename().empty())
+    {
+        devName = parentofParentPath.parent_path().filename() + "_";
+    }
+    if (!parentofParentPath.filename().empty())
+    {
+        devName += parentofParentPath.filename() + "_";
+    }
+    if (!parentPath.filename().empty())
+    {
+        devName += parentPath.filename() + "_";
+    }
+    devName += path.filename();
+    return devName;
+}
 
+/**
+ * @brief get PCIeDeviceList to resp
+ *
+ * @param[in,out]   asyncResp   Async HTTP response.
+ * @param[in]      jsonKeyName
+ */
 inline void
     getPCIeDeviceList(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
                       const nlohmann::json::json_pointer& jsonKeyName)
 {
     static constexpr std::array<std::string_view, 1> pcieDeviceInterface = {
         "xyz.openbmc_project.Inventory.Item.PCIeDevice"};
-    const boost::urls::url pcieDeviceUrl =
-        boost::urls::url("/redfish/v1/Systems/system/PCIeDevices");
 
-    collection_util::getCollectionToKey(
-        asyncResp, pcieDeviceUrl, pcieDeviceInterface,
-        "/xyz/openbmc_project/inventory", jsonKeyName);
+    dbus::utility::getSubTreePaths(
+        "/xyz/openbmc_project/inventory", 0, pcieDeviceInterface,
+        [asyncResp, jsonKeyName](
+            const boost::system::error_code& ec,
+            const dbus::utility::MapperGetSubTreePathsResponse& paths) {
```

Comment:
Should this > 10 line lambda be in a named function?

Reply:
I refactored it.